static/src/javascripts/projects/common/modules/component.js

```diff
@@ -0,0 +1,361 @@
+// @flow
+/* eslint-disable no-underscore-dangle */
+
+import bean from 'bean';
+import bonzo from 'bonzo';
+import fetchJSON from 'lib/fetch-json';
+import qwery from 'qwery';
+
+class ComponentError {
+    constructor(message) {
+        return new Error(`Component: ${message}`);
+    }
+}
+
+class Component {
+    useBem: boolean;
+    templateName: ?string;
+    componentClass: ?string;
+    endpoint: ?string | ?() => string;
+    classes: ?Object;
+    elem: ?HTMLElement;
```

Comment:
If this were:
```js
elem: ?HTMLElement | ?HTMLLinkElement
```
We can refine this elsewhere:
```js
if (typeof this.elem === 'HTMLLinkElement') {
  elem.href = origin + href;
}
```
rather than casting down to `any`.